Source/DafnyCore/ProofDependencyWarnings.cs

```diff
@@ -135,6 +135,10 @@ public class ProofDependencyWarnings {
           lit == true) {
         return false;
       }
+
+      if (assumeDep.Form != AssumptionForm.AssumeStatement && assumeDep.Form != AssumptionForm.AssertStatement) {
```

Comment:
Generally I'm wary of code that is in a generic location (unused specification warnings) and points to specific locations (specific assumption form enum values, which map to AST node types).

An alternative implementation would be to put the knowledge of whether these forms can generate warnings in the various assumption forms, and then this code would be `if (!assumeDep.Form.IsExplicit) { return false; }` 

I've randomly picked the name `IsExplicit`, since I don't know what the condition is upon which you decide which forms should return false here. That you can decide on this name would be another advantage of this code style: it explains better what the property is that makes you decide to return false.

Reply:
Yeah, I think putting the check as a member of the `AssumptionForm` type would be a distinct improvement. Good catch.

Reply:
I ended up putting it in `AssumptionDependency`, which also seems like a reasonable location, since `enum` types in C# can't have members. (You can use extension methods, but that seems messier).